File: linked-list-palindrome/LinkedListPalindrome.java
import java.util.*;

class LinkedListPalindrome {
  public static class Node {
    public int value;
    public Node next;

    public Node(int value) {
      this.value = value;
      this.next = null;
    }
  }

  public boolean LinkedListPalindrome(Node head) {
    Node slowNode = head;
    Node fastNode = head;
    while(fastNode != null && fastNode.next != null) {
      slowNode = slowNode.next;
      fastNode = fastNode.next.next;
    }

    Node reversedSecondHalfNode = reverseLinkedList(slowNode);
    Node firstHalfNode = head;
    while(reversedSecondHalfNode != null) {
      if(reversedSecondHalfNode.value  != firstHalfNode.value) return false;
      reversedSecondHalfNode = reversedSecondHalfNode.next;
      firstHalfNode = firstHalfNode.next;
    }

    return true;
  }

  public Node reverseLinkedList(Node head) {
    Node current = head;
    Node previous = null;
    while(current != null) {
      Node next = current.next;
      current.next = previous;
      previous = current;
      current = next;
    }
    return previous;
  }
}

File: linked-list-palindrome/LinkedListPalindrome.swift
class LinkedListPalindrome {
    class Node {
        var value: Int
        var next: Node?
        init(_ value: Int) {
            self.value = value
            self.next = nil
        }
    }
    // a -> b -> c -> d -> c -> b -> a -> nil
    //                d -> c -> b -> a
    func isPalindrome(_ head: Node?) -> Bool {
        var slow = head
        var fast = head
        while fast != nil, fast?.next != nil {
            slow = slow?.next
            fast = fast?.next?.next
        }

        var reverse = reverseList(slow)
        var secondHalf = head

        while secondHalf != nil, reverse != nil{
            if(secondHalf?.value != reverse?.value) { return false }
            secondHalf = secondHalf?.next
            reverse = reverse?.next
        }
        return true
    }

    func reverseList(_ head: Node?) -> Node? {
        var current = head 
        var previous: Node? = nil
        while(current != nil) {
            var next = current?.next
            current?.next = previous
            previous = current
            current = next
        }
        return previous
    }
}